Playing functions

itivoli opened the pull request:
This is the one of the last steps in allowing the application to generate audio with respect to functions. There are 3 new Nodes: 

**1. AudioGenerator Node,
2. LowerTimeline Node, 
3. Function Node.** 

**AudioGenerator: Main Audio Generation Node.**
This is the one to be interfaced with the rest of the program. This is the one that will control the playing of audio through the Timeline. It houses a LowerTimeline Node that is accessible by the rest of the application, probably specifically only the Timeline UI.

**LowerTimeline: Backend of the Timeline UI. Backbone of AudioGeneration.**
This should interface with the Timeline UI (UpperTimeline?) to fully form the operational Timeline. This node is made up of a list of Function Nodes to be accessed and manipulated in a few ways (all methods and attributes have the necessary comments included). 

**Function: Represents any individual function passed into the program as a Node.**
This node will be the main link between the Function Palette and the IFunctionAST interface that evaluates the functions. Each Function has the capacity to be individually sonified and this is what drives the Timeline and thus the AudioGenerator. It's important to point out that constructing a Function requires a start and end time. As of now, there are test methods inside the class to demonstrate audio generation that need not be present in the final implementation of the Application.

_I also further defined the Function Namespace and created a TimeKeeper class to maintain synchronicity with the processing step of the program._

itivoli committed:
Added Structure of Functions and Timeline

Added a new Timeline class and a Function go along with it. The timeline will be made up of an array of functions and the sequence will be played in order from the streams of each function. Doesn't currently make any sound from the new structures but will soon.

itivoli committed:
Some more Noise changes

itivoli committed:
Now Playing a Full Sequence of Notes

This version plays a full scale (12) of notes.

itivoli committed:
Using process method to play function audio stream

Made an unsuccessful attempt at using the process method to play the audio stream of a function. Committing in case I need to work on this at a different machine.

itivoli committed:
Added Timer

Added Timer Node. Defined the Functions namespace. Moving towards playing audio sequentially from Functions.

itivoli committed:
Timer Added to Function Class

Added a basic implementation of the timer to the function class and each function's process method can now run for a preset duration when triggered.

itivoli committed:
Modified Function Contructor

See Title

itivoli committed:
AudioGenerator Now Functional

The AudioGenerator Node is now functional. Currently running in  a simulated environment to demonstrate playback for audio purposes. Needs to be interfaced with the Timeline UI Node ASAP.

itivoli committed:
Finalizing Changes

Polishing up for PR.

PROMETHIA-27 commented on fnky_parser.dll:
Don't check in the parser.dll; it should be in the .gitignore so I'm not sure how it got in. Old branch maybe?

itivoli replied:
Will do. Maybe I removed a .gitignore I thought I put in accidentally.

PROMETHIA-27 commented on Functions/Functions.cs:
```diff
@@ -0,0 +1,79 @@
+    /// <summary>
+    /// Class for Sycnhronous Timing within the application. 
+    /// </summary>
+    public partial class TimeKeeper {};
```
Do you need this line here? Seems unnecessary

itivoli replied:
I put that there in the same vane that I included `Function` and all the others in the `Functions` namespace. I was under the impression that each file I put the `namespace Functions` declaration in, I should also include in the actual namespace file(?).

PROMETHIA-27 replied:
There is no namespace file. Namespaces are folders, usually. You can just put `namespace X.Y.Z;` at the top of a file and it's fine

PROMETHIA-27 commented on Functions/Functions.cs:
```diff
@@ -0,0 +1,79 @@
+    /// </summary>
+    public partial class TimeKeeper {};
+    
+    namespace Sonification {
```
For organization projects usually have namespaces directly reflect the path to a file. So e.g. `Function` would be defined in `Functions/Sonification/Function.cs`, not `Functions/Function.cs`

itivoli replied:
That makes sense. I put the `Functions/Sonification` in so I could keep everything that dealt with only sound generation in there and I left `Function `out because I figured those nodes could be used within the` Timeline UI` as well, but that seems to be bad reasoning now. Would you suggest we get rid of `Sonification` and have everything in `Functions`?

PROMETHIA-27 replied:
I don't mind the sonification namespace, it's just organized weirdly; You shouldn't have two instances of the keyword `namespace` in one file.

itivoli replied:
I'm down for a reorganization to make it not weird. I didn't know it was bad practice to have the `namespace` keyword occur more than once in a file, I was just going off of this: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/namespaces#:~:text=is%20semantically%20equivalent%20to. Does that mean I should have a new `.cs` file w/ `namespace Functions.Sonfication` in it instead of nesting them like it is right now?

PROMETHIA-27 replied:
The usual rules are:
- Every class goes in its own .cs file with the same name
- Every class's directory path is the same as its namespace. So if the namespace of a class is `Functions.Sonification` then its path is `Functions/Sonification/Class.cs`
- Use the shortform namespace syntax (just a `namespace X.Y.Z;` at the top of the file after your `using`s)
- You basically never need partial classes. We *do* need them for godot stuff because godot generates more stuff to add into them, but for normal purposes classes are defined in one place.

and yeah probably make some new .cs files to split it up.

itivoli replied:
Alright, thanks that makes sense. With the added context, I think I should probably leave the `Functions` namespace to what you've written for the Functions Folder (b/c like you said, namespaces correspond with and are folders), and probably just put everything I did in a new folder and namespace (Sonification I suppose). The last part explains why I got an error when initially trying to remove the partial declaration from the `TimeKeeper` class too, thanks.

PROMETHIA-27 reviewed (commented):
I didn't see a way to take an `IFunctionAST` and produce sound from it, is that possible in this system? It looked like functions were hardcoded songs. Maybe `Function` should have an `IFunctionAST` field to store its AST?

itivoli committed:
Updated Note Sequence Filling, Changed Namespace, Simplified Function Node

In the `Function` Class, simplified the needs to create new function and added in AST support from `IFunctionAST`. Also in the `Function` class, updated the `FillNoteSequence()` method to fill the list that will play the audio using a basic linear mapping to the frequency domain.

PROMETHIA-27 committed:
Merge branch 'main' into Playing-Functions

PROMETHIA-27 committed:
Merge branch 'Playing-Functions' of https://github.com/PROMETHIA-27/SoundOfFunctions into Playing-Functions

PROMETHIA-27 committed:
remove dll

PROMETHIA-27 reviewed (approved):
Updated to latest main and removed .dll file. Looks good as is!

PROMETHIA-27 committed:
Merge branch 'main' into Playing-Functions

itivoli committed:
AudioGenerator Playing Some Functions

Using a simple (too simple) linear mapping, the AudioGenerator Node can now comfortably play any function that is fully continuous over the default interval except constant functions. Thus, those are two things to fix.

MW-0317 committed:
Merge branch 'main' into Playing-Functions

itivoli committed:
Fixed Function Oversight

The Audio Streams of each function upon completion, though empty, would continue "playing". That is no longer the case. Fixed some other small things and made some slight improvements. For future integration with the larger application.

itivoli committed:
Merge branch 'Playing-Functions' of https://github.com/F-nk-Sound/SoundOfFunctions into Playing-Functions
